File: code/body.py
from math import sqrt

G = 6.673e-11  # gravitational constant


class Body:
    # cartesian positions
    rx: float
    ry: float

    # velocity components
    vx: float
    vy: float

    # force components
    fx: float = 0.0
    fy: float = 0.0

    mass: float

    def __init__(self, rx: float, ry: float,
                 vx: float, vy: float,
                 mass: float):
        self.rx = rx
        self.ry = ry
        self.vx = vx
        self.vy = vy
        self.mass = mass

    def __str__(self):
        return "{}, {}, {}, {}, {}".format(
            self.rx, self.ry, self.vx, self.vy, self.mass
        )

    def update(self, dt: float):
        """
        Update the velocity and position using a time step dt
        """
        self.vx += dt * self.fx / self.mass
        self.vy += dt * self.fy / self.mass
        self.rx += dt * self.vx
        self.ry += dt * self.vy

    def distance_to(self, other_body: 'Body') -> float:
        """'
        Return the distance between two bodies
        """
        dx = self.rx - other_body.rx
        dy = self.ry - other_body.ry
        return sqrt(dx*dx + dy*dy)

    def reset_force(self):
        """
        Reset the force to 0 for the next iteration
        """
        self.fx = 0.0
        self.fy = 0.0

    def add_force(self, other_body: 'Body'):
        """
        Compute the net force acting between this body and
        other_body and add to the net force acting on this
        body
        """
        # softening parameter (just to avoid infinities)
        eps = 3E4
        dx = other_body.rx - self.rx
        dy = other_body.ry - self.ry
        dist = sqrt(dx*dx + dy*dy)
        F = (G * self.mass * other_body.mass) / (dist*dist + eps*eps)
        self.fx = self.fx + F * dx / dist
        self.fy = self.fy + F * dy / dist

    def is_in(self, quad: 'Quad'):
        return quad.contains(self.rx, self.ry)

File: NBody/Simulation/Body.swift
//
//  Body.swift
//  NBody
//
//  Created by Daniel Jilg on 14.04.18.
//  Copyright © 2018 breakthesystem. All rights reserved.
//

import AppKit
import Foundation

/// Any object a universe keeps track of
class UniverseObject: CustomStringConvertible {
    var tickNumber: Int = 0
    var name: String
    lazy var color: NSColor = { return NSColor.random() }()

    init(tickNumber: Int, name: String) {
        self.tickNumber = tickNumber
        self.name = name
    }

    var description: String {
        return name
    }

    /// Update the UniverseObject between calculated frames
    func update(timeSteps: Double) {
    }
}

/// Any *physical* object a universe keeps track of
class Body: UniverseObject {
    /// The body's current position in the universe.
    var position: Position

    /// The body's travel direction and speed.
    /// Will be added to the body's position at the beginning of each step.
    var direction: Vector

    /// The sum of acting forces acting on this body.
    /// Will be applied to the direction vector at the end of each calculation step,
    var force: Vector

    init(position: Position, direction: Vector, name: String = "Unnamed Body", tickNumber: Int = 0) {
        self.position = position
        self.direction = direction
        self.force = Vector(x: 0, y: 0)

        super.init(tickNumber: tickNumber, name: name)
    }

    /// Set the sum of acting forces to zero
    func resetForce() {
        force = Vector(x: 0, y: 0)
    }

    func move(for timeSteps: Double) {
        position = position + direction * timeSteps
    }
}

/// A large physical object with considerable mass
class Planetoid: Body {
    var mass: Double

    init(position: Position, direction: Vector, name: String = "Unnamed Planetoid", tickNumber: Int = 0, mass: Double = 7.34767309e22) {
        self.mass = mass

        super.init(position: position, direction: direction, name: name, tickNumber: tickNumber)
    }


    /// Update the velocity and position
    // TODO: This should work with a body as well
    override func update(timeSteps: Double) {
        let timesForce = force * timeSteps
        let massAdjustedTimesForce = timesForce / mass
        direction = direction + massAdjustedTimesForce
        move(for: timeSteps)
    }

    /// Compute the net force acting between this body and otherBody and add to the net force acting on this body
    // TODO: This should work between a body and a planetoid as well
    func accelerate(with otherPlanetoid: Planetoid) {
        guard self !== otherPlanetoid else { return }
        guard otherPlanetoid.name == "Sün" else { return }

        // softening parameter (just to avoid infinities)
        let eps = 3E4

        let distance = position.distance(to: otherPlanetoid.position)
        let distanceX = otherPlanetoid.position.x - position.x
        let distanceY = otherPlanetoid.position.y - position.y
        let actingForce = (G * mass * otherPlanetoid.mass) / (distance.squared + eps.squared)

        force = Vector(
            x: direction.x + actingForce * distanceX / distance,
            y: direction.y + actingForce * distanceY / distance
        )
    }
}
